### ISSUE TITLE
"The route is not valid" error is not helpful

### ISSUE THREAD

**Issue by musicpanda** [PrestaShop/PrestaShop#36687]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add attachments

When you get a "The route is not valid" error it only shows the route, not the rule for which the route is for. 

In the case I was investigating some mysterious "tag_rule" was missing. I am still investigating that - possible a remainder of a deleted module. It would be a small effort to mention both the route and what it is for.

The relevant code is in
/src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/TrafficSeo/Meta/MetaSettingsFormDataProvider.php

In the function validateUrlSchema() it says:
`        foreach ($configuration as $routeId => $rule) {
            if (!$this->routeValidator->isRoutePattern($rule)) {
                $patternErrors[] = $this->translator->trans(
                  'The route %routeRule% is not valid',
                  [
                      '%routeRule%' => htmlspecialchars($rule),
                  ],
                  'Admin.Shopparameters.Feature'
                );
            }`

### Expected behavior

Tell for what the rule is.

### Steps to reproduce

Create a faulty rule and watch the error.

### PrestaShop version(s) where the bug happened

8.1.4

### PHP version(s) where the bug happened

8.1

**Comment by florine2623** [comment]
Hello @musicpanda ,

Are you talking about Cart rules ? Catalog Price rules ?

Could you explicit in more details the steps to reproduce your bug please ?

Thanks, waiting for your feedback.

**Comment by musicpanda** [comment]
Hello @florine2623 : it should be obvious that these are not cart rules. Cart rules are not routes. I assume a certain level of  background knowledge here.

In the Prestashop backoffice SEO&Urls page you have the routes. They determine how "pretty urls" look like. They look like this:
![routes](https://github.com/user-attachments/assets/7f26de5a-74c9-4b49-8a71-6d3a67a968c1)

When those fields are empty or invalid you get an error message: "The route is not valid" with behind the word "route" the rejected string - if there was any. However, it doesn't mention which of the eight routes was invalid. You are supposed to find that out yourself. 

In the shop I worked on a module (zhproducttag) had added a ninth route with the name tag_rule. For some reason that field had become empty. The result was that I was no longer able to regenerate .htaccess as that produced the "The route is no valid" error.

Of course it was not obvious from this error message that the problem was with this module and it took me a considerable amount of debugging time to find out.

If you look on the forum the "The route is not valid" problems are quite common. Some people report fixing it after removing some (often not mentioned) module. Others probably never solve it. 

This kind of problems would be a lot easier to solve when the route name was mentioned in the error message.

**Comment by florine2623** [comment]
Hello @musicpanda ,

I can't reproduce your issue.

I do not get the error "The route is not valid".

I have this type of error, if I'm not using the right keyword of if a mandatory keyword is not used :
<img width="1211" alt="Screenshot 2024-08-09 at 15 09 43" src="https://github.com/user-attachments/assets/e2e180d3-d2d4-4644-9189-3dfaa93cd144">

Do you mean that you want this above error message to be clearer ?

Plus, if my field is empty, I can't save my form.

<img width="836" alt="Screenshot 2024-08-09 at 15 10 56" src="https://github.com/user-attachments/assets/e9a17be5-3a66-4a1f-9e6d-6ec321cb4d92">

**Comment by musicpanda** [comment]
Hello @florine2623 ,

You are just trolling me. Time and again you deliberately misunderstand me and then you claim that you cannot reproduce it.

If you want to contribute, please read what I have written carefully before you react.

**Comment by Hlavtox** [comment]
@musicpanda Hello, Florine tried to reproduce your issue the best she can, trying bad keywords or making it empty. You can see that she did not get the error message you talk about.

If you want her to reproduce the bug, provide a faulty module and EXACT steps to get THE SAME ERROR MESSAGE.

-----------

> I assume a certain level of background knowledge here.
> You are just trolling me. Time and again you deliberately misunderstand me and then you claim that you cannot reproduce it.

Please stop behaving like a dickhead, you have been already banned one time and you will be banned second time. This is not a channel where you can complain about a software that you got for free and insult everyone you talk to. You don't like it? It has bugs? Fix it yourself.

**Comment by musicpanda** [comment]
@Hlavtox It is you who are misbehaving with words like dickhead. I don't use such terms.

This error message in the code is clearly not helpful. I am puzzled why anyone would even want to reproduce the problem. Don't you have basic principles that error messages should be helpful?

A simple search on the forum will tell you that quite a few people have the problem I mentioned. A clearer error message would have helped them.

I mentioned the module name already and for copyright reasons I cannot include the code. The only thing I can add are a few more details:
 - it concerns the zhproducttag module by Zh-Soft
 - If you configure the module you will see the "Route to tags" path. It is an obligatory field and I have no idea why it got empty in my case. As it only becomes visible when you try to recreate .htaccess the problem may have been around for some time. Unfortunately software is not perfect and sometimes fields get an invalid value.

**Comment by jolelievre** [comment]
Hi @musicpanda,

to reproduce the error we need a module that adds a similar route possibility, with similar constraints in its validation. This is what Florin needs to reproduce the bug, so I understand you can't share the `zhproducttag` module completely, but is it possible for you to create an empty module that would just adds this specific route? So we can reproduce the bug and think about a proper improvement.

Also, I don't think this is a bug, the validator behaves as it should it's not responsible for the lack of info in the error message. This is more an improvement that could be added in the next v9 version.

**Comment by musicpanda** [comment]
Hi @jolelievre ,

I give up. This is hopeless. When I look at the configuration table of this shop I see no "ROUTE" variable at all. I don't know whether they were never made or somehow deleted. Even giving you the whole module would not bring any light to what has happened in this shop and whether the module was even responsible. I am puzzled why Prestashop people want to waste their time. 

I disagree with you that this is not a bug. Faulty error messages are major bugs. And Prestashop has quite a few of them. It is costing the people working with Prestashop code lots of time and headaches.

**Comment by matks** [comment]
The problem is here:

> I assume a certain level of background knowledge here.

Don't assume a certain level of background knowledge. Please write down comments as if you're talking to someone who started using PrestaShop 2 weeks ago. This software is big, it has been written by hundred of people who came and go, and most of the people who are working on the project today have joined the boat somewhere along the way and have gained partial knowledge of the software by trial and experiment. I think nobody today has a 100% knowledge of the PrestaShop software. Don't assume that because we are part of the project team we have a certain level of background knowledge about everything.

I have worked with PrestaShop since 2018 and I think after 6 years of tinkering with it I only know 50 or 60% of it.

This is why we write `If you want her to reproduce the bug, provide a faulty module and EXACT steps to get THE SAME ERROR MESSAGE.` . It's not to annoy you. It is because the easiest way to communicate a problem to someone is to have him experience it, and to do that we need a list of "step 1, do (...), step 2, do (...), step 3 ..."

> I disagree with you that this is not a bug. Faulty error messages are major bugs.

Definition of bug is personal. For some people a bug is "the software was intended to perform the behavior (...) and the behavior (...) does not happen". So if the 'Pretty URL' feature was designed with this error message to be written, I agree it is not a bug: the software behaves as it was designed, it follows the specifications, and then you are saying "I think the specifications were not user friendly and I suggest to improve the specificiations and match the software behavior to obtain a better user experience".

**Comment by musicpanda** [comment]
@matks ,
I get the impression that you want a step-by-step description not to understand the problem but to make it possible that even people who have never seen Prestashop can solve it. Step-by-step should even solve the problem that someone cannot distinguish a route from a cart rule.

The problem is that not all problems can so easily be caught. That applies to the problem here too. The problem was discovered long after the installation of the module and it may have been around for a long time. As the field is obligatory in the module **it is not supposed to happen** that it is empty. So no, I cannot give step-by-step instructions how to arrive at this situation. I can only describe the problem and say that I would have solved it much easier if Prestashop's error message had been more helpful.

**Comment by matks** [comment]
> @matks , I get the impression that you want a step-by-step description not to understand the problem but to make it possible that even people who have never seen Prestashop can solve it

Almost there, but no 100%

I want a step-by-step to make it possible that even people who have never seen Prestashop can _reproduce_ it.

Reproduce is not solving 😄 else the world would be much more friendly

> So no, I cannot give step-by-step instructions how to arrive at this situation. I can only describe the problem and say that I would have solved it much easier if Prestashop's error message had been more helpful.

Then you are not reporting a BUG (something we can reproduce and label) but a DEFECT (also known as a TECHNICAL FACT), similarly to https://github.com/PrestaShop/PrestaShop/issues/24251

And I already answered that here https://github.com/PrestaShop/PrestaShop/issues/24251#issuecomment-829949238

I know you are trying to help

I know you are trying to give useful things to us

And I guess the reaction you expected from us was: "Oh? Interesting. Thanks for telling us, we were not aware of that. Thanks to this information we'll be able to do something and make the software better".

I am sorry, PrestaShop is not a project where what I just wrote will happen. Truth is: everybody here is over-busy with its own priorities and topics. We all have already a big pile of problems to solve with PrestaShop. So we cannot welcome the things you give us and "take care of it": we already are under heavy workload.

You can continue reporting defects using [github discussions](https://github.com/PrestaShop/PrestaShop/discussions). They might be useful: someone could read them and learn from it. Someone with the same problem as you will benefit from the context you can provide. But we will not take the defects you have found in the software - unless they are critical - and deal with it.

**Comment by musicpanda** [comment]
@matks 
That sounds to me like a very sick way to manage a software project.

It declares the kind of problem we discuss here unsolvable. The way to reproduce this problem in my shop is just to regenerate .htaccess and you will see the message at the top of your page. But when you write that here on Github you will get the reaction that someone could not reproduce it in his shop and the issue is closed. 

This happens almost daily here on Github. A gigantic waste of human effort. 

It is not hard to understand that there must be some underlying cause that makes a problem appear on one shop and not on another. It may be a module or some database consistency. It can be a very specific sequence of actions on a product or an order. In the absence of a clear error message the only way to find the origin of such problems is hardcore debugging. 

What seldom works is asking people endless details about their shop. It gives the reporter of the problem loads of work and it is unlikely that anyone will do anything with it. It reminds me of the classic helpdesk trick of asking people lots of questions in the hope that they won't come back.

This "The route is not valid" message is typical of dysfunctional error messages. Not only refuses it to tell you which route causes the problem. It is also written in such a way that nobody understands why. The faulty route value appears between the words 'route' and 'is' but I doubt that anyone who sees this message understands that the problem is an empty route value.

> So we cannot welcome the things you give us and "take care of it": we already are under heavy workload.
I beg to see that a bit different. When I report a problem I have done the hard work. I have analyzed the problem and I have formulated how it can be solved. Sure, I can't be bothered with Prestashop's coding guidelines and the blackbox that is called Github. But as there seem to be lots of people who feel at home in that environment I don't believe I am burdening them that much. 

As for "github discussions", my preferred place is the Prestashop forum as that is the place where problems are solved. I still find it a waste that Prestashop team has discarded its interface there.

**Comment by matks** [comment]
> That sounds to me like a very sick way to manage a software project.

When you write things like this, it simply makes the person on the other side stop listening 😞 

I won't answer anymore in this conversation, I think nothing else positive will come out of it. I explained the things as they are, I don't need your opinion on that, especially when we are the ones doing the work. You're basically looking at people, doing some work for you for free (nobody made you pay for all the work we did on this issue) and you are criticizing the outcome.

I also close the issue: it's pretty clear it'll never become a reproduceable bug report.

If you don't understand my reaction, read [this article](https://snarky.ca/the-social-contract-of-open-source/).
